cli/src/evmsigner.ts

```diff
@@ -0,0 +1,205 @@
+  Provider,
+  TransactionRequest,
+} from 'ethers';
+import { NonceManager, Wallet } from 'ethers';
+
+export async function getEvmSigner(
+  rpc: Provider,
+  key: string | EthersSigner,
+  opts?: {
+    maxGasLimit?: bigint;
+    chain?: EvmChains;
+    debug?: boolean;
+  },
+): Promise<Signer> {
+  const signer: EthersSigner =
+    typeof key === 'string' ? new Wallet(key, rpc) : key;
+
+  const chain = opts?.chain ?? (await EvmPlatform.chainFromRpc(rpc))[1];
+  const managedSigner = new NonceManager(signer);
+
+  if (managedSigner.provider === null) {
+    try {
+      managedSigner.connect(rpc);
+    } catch (e) {
+      console.error('Cannot connect to network for signer', e);
+    }
+  }
+
+  return new EvmNativeSigner(
+    chain,
+    await signer.getAddress(),
+    managedSigner,
+    opts,
+  );
+}
+
+// Get a SignOnlySigner for the EVM platform
+export async function getEvmSignerForKey(
+  rpc: Provider,
+  privateKey: string,
+): Promise<Signer> {
+  return getEvmSigner(rpc, privateKey);
+}
+
+// Get a SignOnlySigner for the EVM platform
+export async function getEvmSignerForSigner(
+  signer: EthersSigner,
+): Promise<Signer> {
+  if (!signer.provider) throw new Error('Signer must have a provider');
+  return getEvmSigner(signer.provider!, signer, {});
+}
+
+export class EvmNativeSigner<N extends Network, C extends EvmChains = EvmChains>
+  extends PlatformNativeSigner<EthersSigner, N, C>
+  implements SignOnlySigner<N, C>
+{
+  constructor(
+    _chain: C,
+    _address: string,
+    _signer: EthersSigner,
+    readonly opts?: { maxGasLimit?: bigint; debug?: boolean },
+  ) {
+    super(_chain, _address, _signer);
+  }
+
+  chain(): C {
+    return this._chain;
+  }
+
+  address(): string {
+    return this._address;
+  }
+
+  async sign(tx: UnsignedTransaction<N, C>[]): Promise<SignedTx[]> {
+    const chain = this.chain();
+
+    const signed = [];
+
+    // default gas limit
+    const gasLimit = chain === 'ArbitrumSepolia'
+      ? 4_000_000n
+      : this.opts?.maxGasLimit ?? 500_000n;
+
+
+    // TODO: DIFF STARTS HERE
+
+    let gasPrice = 200_000_000_000n; // 200gwei
+    let maxFeePerGas = 6_000_000_000n; // 6gwei
+    let maxPriorityFeePerGas = 1000_000_000n; // 1gwei
+
+    // Celo does not support this call
+    if (chain !== 'Celo') {
+      const feeData = await this._signer.provider!.getFeeData();
+      gasPrice = feeData.gasPrice ?? gasPrice;
+      maxFeePerGas = feeData.maxFeePerGas ?? maxFeePerGas;
+      maxPriorityFeePerGas =
+        feeData.maxPriorityFeePerGas ?? maxPriorityFeePerGas;
+    }
+
+    // Oasis throws malformed errors unless we
+    // set it to use legacy transaction parameters
+    const gasOpts =
+      chain === 'Oasis'
+        ? {
+            gasLimit,
+            gasPrice: gasPrice,
+            // Hardcode type
+            type: 0,
+          }
+        : {
+            gasPrice,
+            maxFeePerGas,
+            maxPriorityFeePerGas,
+            gasLimit,
+          };
+
```

Comment:
mostly feedback:
Generally as we deploy to mainy evms we find that these kind of special cases for gas price happen in many chains, so this is something that will probably be modified often... I think it makes sense to move this into a separate function so that anyone using the cli can modify it quickly without fear of breaking something else

Reply:
thanks for the feedback, we can iterate on this in the future as needed 🙏